runtime/oti/j9nonbuilder.h

```diff
@@ -2827,10 +2827,19 @@ typedef struct J9Object {
 #define OBJECT_HEADER_LOCK_INFLATED  1
 #define OBJECT_HEADER_LOCK_FLC  2
 #define OBJECT_HEADER_LOCK_RESERVED  4
-#define OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT  8
+#define OBJECT_HEADER_LOCK_LEARNING  8
+#define OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT  0x10
 #define OBJECT_HEADER_LOCK_LAST_RECURSION_BIT  0x80
-#define OBJECT_HEADER_LOCK_RECURSION_MASK  0xF8
-#define OBJECT_HEADER_LOCK_RECURSION_OFFSET  3
```

Comment:
This old value must be used within `ObjectMonitor_V1` otherwise old core files will be misinterpreted. Perhaps we should keep the old macros and use new in `ObjectMonitor_V2` and elsewhere.

Reply:
Would it make sense to just use newly-named constants and leave the old ones alone (documented so they don't get deleted later!)?

Reply:
There are 3 defines that need new values: `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT`, `OBJECT_HEADER_LOCK_RECURSION_MASK` and `OBJECT_HEADER_LOCK_RECURSION_OFFSET`.

I understand the justification for leaving in `OBJECT_HEADER_LOCK_RECURSION_OFFSET` for use in `ObjectMonitor_V1`. But, I don't really understand why old versions of the other two are still needed if only the new values are used. Is there a particular reason you feel like it might be useful  to keep them around? If not, it seems like it is just adding more complications without any real benefit.

Reply:
Any constants used by `ObjectMonitor_V1` must not change else we will be unable to properly interpret old core files (where `ALG_OBJECT_MONITOR_VERSION` is 0). We can either add new constants to be used for the new implementation (with comments warning against new uses of the old constants) or inline the old constant values into `ObjectMonitor_V1`. Graham suggested the former plan and I agree it is the better approach.

Reply:
I understand that `ObjectMonitor_V1` needs the older values. However, it only looks like `OBJECT_HEADER_LOCK_RECURSION_OFFSET` is used inside `ObjectMonitor_V1`. `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT` and `OBJECT_HEADER_LOCK_RECURSION_MASK` are not used in `ObjectMonitor_V1`.

Reply:
Ok, `OBJECT_HEADER_LOCK_RECURSION_OFFSET` is the only constant used by `ObjectMonitor_V1` that's changing now. If other used constants change in the future, that will break support for old core files. I suggest we need a comment that warns about changing the constants that must not change without fixing `ObjectMonitor_V1` accordingly.

I think the constants for the new implementation should use a different prefix: I suggest `OBJECT_HEADER_LOCK_V2_`.

Reply:
Okay then I will make the following changes:
New `OBJECT_HEADER_LOCK_FIRST_RECURSION_BIT` -> `OBJECT_HEADER_LOCK_V2_FIRST_RECURSION_BIT`
New `OBJECT_HEADER_LOCK_RECURSION_MASK` -> `OBJECT_HEADER_LOCK_V2_RECURSION_MASK`
New `OBJECT_HEADER_LOCK_RECURSION_OFFSET` -> `OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET`

I will also add a comment regarding the connection to `ObjectMonitor` java files.

Reply:
I've added the changes in the latest commit.